### ISSUE TITLE
Orquesta - Support output-on-error semantics

### ISSUE THREAD

**Issue by nmaludy** [StackStorm/st2#4336]
##### SUMMARY

In Mistral workflows there was an option for an `output-on-error` stanza. This block allowed workflow authors to return data in the event of an error within the workflow. We use this feature internally do things like:

- Returning meaningful error messages from workflows
- Returning the data that caused the error from a workflow
- Returning a list of tasks that completed and those that failed within a variable (used for emails, ticketing systems and dashboards)

I would be open to a different mechanism (does not need to be `output-on-error`). But some way to replicate this functionality would be great!

##### ISSUE TYPE
 - Feature Request

##### STACKSTORM VERSION
```shell
$ st2 --version
st2 2.8.1, on Python 2.7.5
```

##### OS / ENVIRONMENT / INSTALL METHOD

```
OS = CentOS 7.5
Install method = puppet-st2
```

##### Example Mistral workflow

```yaml
encore.preprovision:
  type: direct
  input:
    - domain_type
    - dns_domain
    - hostname
    - task_list: []
    - vm_os

  output:
    ip_address: "{{ _.ip_address | default(None) }}"
    task_list: "{{ _.task_list }}"
    vm_dns_records: "{{ _.dns_records | default(None) }}"
    vm_os_name: "{{ _.vm_os_name | default(None) }}"
  output-on-error:
    task_list: "{{ _.task_list }}"
    vm_dns_records: "{{ _.dns_records | default(None) }}"

  tasks:

...

    validate_ipam_hostname_unique:
      action: encore.mm_check_hostname
      input:
        hostname: "{{ _.hostname }}"
        dns_domain: "{{ _.dns_domain }}"
      publish:
        task_list: "{{ _.task_list + [{'name': 'validate_ipam_hostname_unique', 'status': 'SUCCESS'}] }}"
      publish-on-error:
        task_list: "{{ _.task_list + [{'name': 'validate_ipam_hostname_unique', 'status': 'ERROR', 'err_reason': 'Host already exists with name: ' + _.hostname + '.' + _.dns_domain }] }}"
      on-success:
        - validate_ad_computer_unique: "{{ _.vm_os.type == 'windows' and _.domain_type.lower() == 'domain' }}"
        - build_dns_records: "{{ _.vm_os.type != 'windows' or _.domain_type.lower() != 'domain' }}"
      on-error:
        - send_error_email
...
```

**Comment by m4dcoder** [comment]
Here's another variation. We can add a log entry to be included in the errors of orquesta workflow result.  Orquesta workflow currently has list of errors and output in the result.

```
version: 1.0

description: A workflow demonstrating fail command.

tasks:
  task1:
    action: core.local
    input:
      cmd: '>&2 echo "boom!"; exit 1'
    next:
      - when: <% failed() %>
        publish:
          - err_msg: 'task1 failed'
          - extra_data: <% result().stderr %>
        log:
          - message: <% ctx(err_msg) %>
            result: <% ctx(extra_data) %>
        do: fail
```

**Comment by m4dcoder** [comment]
And the action execution would look something like ```id: 5b996bfe0a08a4187eca165c
action.ref: foobar
status: failed
start_timestamp: Wed, 12 Sep 2018 19:41:50 UTC
end_timestamp: Wed, 12 Sep 2018 19:41:51 UTC
result: 
  log:
    - message: task1 failed
      result: boom!
      task_id: task1
  errors:
    - message: Execution failed. See result for details.
      result:
        failed: true
        return_code: 1
        stderr: boom!
        stdout: ''
        succeeded: false
      task_id: task1
  output: null```

**Comment by m4dcoder** [comment]
@dzimine @bigmstone I would like your input on the design above. I kinda prefer the latter.

**Comment by cognifloyd** [comment]
@m4dcoder: In @nmaludy's example, the `output-on-error` is at the workflow level, not the task level. It's a catch-all, what to do if something in the workflow itself fails, not if an individual task fails.
https://specs.openstack.org/openstack/mistral-specs/specs/ocata/approved/publish-on-failure.html

(@nmaludy I haven't actually used this before - is that right?)

**Comment by m4dcoder** [comment]
@cognifloyd IMO, if the workflow fails, then there's no output (or return value) from the execution. Think of this as a function call. If the function fails, it'll return errors or print errors to console, but it doesn't have a return value.  Therefore, if the workflow author wants to output something for troubleshooting, the option for a `log` command that I propose above let's user write messages and any extras. Since the placement of the `log` command is coupled with the condition of the task execution, it allows for user to determine when to write the log entries.  Please remember that an error could also be a like HTTP where the execution is successful but with an error code. In this case, output-on-error or publish-on-error semantics won't be able to catch them.  Then finally, on completion of the workflow execution, regardless of success or failure, we will return the log entries in the workflow result in a log section like we do for errors.

**Comment by m4dcoder** [comment]
@cognifloyd also, orquesta handles unexpected exceptions at the workflow level and write error message under `errors` in result. I also have seen what @nmaludy do in his workflows. So he can use `log` to log progress with any extra data and then if workflow fails unexpectedly, the log progress + data + errors will be in the workflow result.

**Comment by jeking3** [comment]
I'd prefer to have the option to do output-on-error with Orquesta too, or perhaps just to always do it.  In one example I have a workflow that runs a remote command.  If the remote command fails I still want to be able to get the stdout and stderr and exit_code of the remote command, just like when the workflow succeeds.  What you are suggesting is that now I need conditional logic to get stdout from my remote command:
```
    if my_workflow.status == 'succeeded':
        stdout = my_workflow.result.output.stdout
    else:
        stdout = my_workflow.log.extra.stdout
```
That's really sub-optimal for the use case where you always want output, regardless of whether the workflow succeeds or fails.  I assume that's why mistral supports it.

**Comment by m4dcoder** [comment]
You don't have to fail the workflow in your example. Just let the workflow succeeds and write the error code and stdout to the workflow output. Think of your use case as HTTP calls.

**Comment by m4dcoder** [comment]
@jeking3 We are not trying to duplicate Mistral exactly and we are very selective what we introduced into Orquesta. We will revisit features and requests from time to time.

**Comment by emptywee** [comment]
To me, not giving users the option to publish output on workflow failure is counter intuitive and goes against common sense. Similar to running an executable and not getting any output at all if it fails for any reason, not even an exit code. 

Failed workflows should let users publish any arbitrary variables with all context available. 
Even when implementing HTTP calls any framework lets you respond with non-200 status code and any headers and body added to it.

I can elaborate more on all this, but making Orquesta limited in this regards makes it way harder to migrate from Mistral. 

Output from workflows is imperative for chatops aliases, when you need to pass various results from workflows to users. Re-writing mistral workflows so that instead of failing we go to a core.noop task where we publish variables and only after that they go to output is not feasible, let alone it masks their logical end of life in the UI and adds more tasks to workflows where not needed.

**Comment by m4dcoder** [comment]
@emptywee Please review the entire issue.  We hear that users want to have some sort of output in the result to be consumed.  We currently don't support `output-on-error` schema and I am not backing down from this.  However, we proposed the `log` schema in the task transition. Whatever that is logged will be available in the result which will be structured as the following: `{'log': [...], 'errors': [...], and 'output': {...}}`.  This should be very useful because now users can log additional message or data that will be included in the result regardless of workflow state.  Each log entries will have specific task id and task transition id to help with identifying source of the message.

**Comment by m4dcoder** [comment]
So given the contention and misunderstanding on the subject, what I will be doing is to let orquesta process the `output` section when the workflow completes (regardless of success or failure). The values will be evaluated based on the final workflow context at time of workflow completion. If there is an evaluation failure on a key value pair, then the key value pair for the output will be skipped and an error entry will be appended in the result describing the evaluation error.  I think this is sufficient in addition with the `log` schema in task transition. Uers should be able to take advantage with the change in `output` and `log` for troubleshooting and other advanced uses.

**Comment by m4dcoder** [comment]
@emptywee please review my comments just above that we plan to change when we process `output`.  that should satisfy the contention here.

**Comment by emptywee** [comment]
@m4dcoder I did, thanks.

**Comment by m4dcoder** [comment]
Workflow output addressed at https://github.com/StackStorm/orquesta/pull/101

**Comment by shusugmt** [comment]
I have tested the fix with the package from #4436 and LGTM. I could migrate our current mistral workflow which were dependent on this behavior, to orquesta and all runs as expected. Thanks for the rapid fix!!
